bgpv1: Consolidate CRD API to follow K8s API Conventions

rastislavs opened the pull request:
This PR aims to consolidate the BGP control plane k8s API to follow the [K8s API Conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md).

Although it modifies some filed names, it does it only for fields that have been added recently (not yet released in any Cilium version) as part of:
 - https://github.com/cilium/cilium/pull/25408
 - https://github.com/cilium/cilium/pull/25660
 - https://github.com/cilium/cilium/pull/25708
 - https://github.com/cilium/cilium/pull/25809

Already released API fields are modified only in a backward compatible way (e.g. changed internal type to a pointer, but no type change in the generated `CustomResourceDefinition`). The only changes of already released API in the generated [ciliumbgppeeringpolicies.yaml](https://github.com/cilium/cilium/pull/26040/files#diff-dddec5a0dbbf46faca6e3ea84b23aefe87d43e6ced346582f5d86d362936f204) are:
- `exportPodCIDR`: added default value (`false`, so not changing the default behaviour)
- `localASN` + `peerASN`: added `format` annotation, but without the change of the underlying `type` 

The changes are described in more detail in individual commits. 

Although the diff looks pretty big, most of the changes are in the unit / component tests, where the API is actually used.

Fixes https://github.com/cilium/cilium/issues/25871

rastislavs committed:
bgpv1: Represent duration fields as integer fields in CRD API

Modifies the type of recently added CRD fields of type
`meta/v1.Duration` to integer type and their name to follow
the K8s API Conventions rule:

"Duration fields must be represented as integer fields with
units being part of the field name (e.g. leaseDurationSeconds).
We don't use Duration in the API since that would require clients
to implement go-compatible parsing."

Also enables defaulting for these fields on apiserver by using
the kubebuilder:default tags (for static defaults).

This includes the KeepaliveTimeSeconds field, which was defaulted
at runtime to 1/3 of HoldTimeSeconds. As per K8s API Conventions,
static defaults are preferred. It also makes the value more
transparent to the user.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs committed:
bgpv1: Move constants with default values to the API package

This allows the default value to be versioned with the API.
It also provides consistency with the kubebuilder tags that
define the default values - both are kept in the same source file.

Also moves timers validation logic to API package to co-locate
the validation logic for API constraints that cannot be
expressed with kubebuilder markers with the API itself.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs committed:
bgpv1: Consolidate CRD field types to follow K8s API Conventions

Modifies the types of CRD fileds to follow  the K8s API
Conventions rules:
- "Use pointer types for all optional fields that do not have
built-in nil value"
- "Do not use unsigned integers, due to inconsistent support
across languages and libraries"
- "All public integer fields MUST use the Go (u)int32 or Go
 (u)int64 types, not (u)int"

Also changes the types of the existing ASN CRD fileds to int64
to satisfy the above rules.

Defaults ExportPodCIDR field of CiliumBGPVirtualRouter to false
explicitly for consistency with other optional fields.
As the default behavior remains the same, this is
a backward-compatible change.

Pointer types can be directly dereferenced in the reconcilers code.
To allow that we provide defaulting methods for our API types,
that can be used from the controllers / unit tests.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

rastislavs commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -431,62 +429,51 @@ func (c *Controller) FullWithdrawal(ctx context.Context) {
 }
 
 // applyPolicyDefaults applies default values on the CiliumBGPPeeringPolicy.
+// This defaulting is normally done by kube-apiserver (based on CRD default markers), but ensured here for non-k8s environments, tests, etc.
```
Actually, I am not sure if we still really need this runtime defaulting. Whole `applyPolicyDefaults` could be removed if we make sure we deference pointers carefully with applying defaults (see e.g. `pointer.Int32Deref(n.HoldTimeSeconds, v2alpha1api.DefaultBGPHoldTimeSeconds)`).

harsimran-pabla replied:
+1 to use `Int32Deref`, this method with a bunch of `if ( ptr == nil || *ptr ==0 ) {...}` will grow as more optional fields are added.

YutaroHayakawa replied:
> if we make sure we deference pointers carefully with applying defaults

Sorry, I didn't get what this means. Does this mean when we dereference the optional fields in the reconcilers, we need to use `Int32Deref`? If so, can't we set the default value at the very beginning and allow reconcilers to dereference without the risk of nil pointer dereference or having an invalid value?

rastislavs replied:
Yeah, I think you understood correctly @YutaroHayakawa . We can do one of the following:

1. always deference with care, e.g. via `pointer.Int32Deref` and we could remove `applyPolicyDefaults()`
2. fully rely on `applyPolicyDefaults()` and always deference directly like `*this`
3. do both  `pointer.Int32Deref` and `applyPolicyDefaults()`, just in case

At the moment we are at the option 3) which I guess is the least ideal?

rastislavs replied:
We discussed this a bit more with @YutaroHayakawa and decided to allow direct pointer dereferences. Apart from making it easier for reconciler developers, it is also safer if we always default before passing the API structs to the reconcilers (in case some developer would forget a nil check / safe dereference). In https://github.com/cilium/cilium/pull/26040/commits/e7faf3704e65edfd32559863310722a417043ad0 I cleaned this up and moved the defaulting logic to the API package, so that it can be used from other packages as well (was needed for unit tests).

harsimran-pabla replied:
Sounds good to me.

rastislavs commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -112,33 +123,32 @@ type CiliumBGPNeighbor struct {
-	// +kubebuilder:validation:Format=duration
-	KeepAliveTime metav1.Duration `json:"keepAliveTime,omitempty"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default=30
```
I thought it would be nicer to default `keepAliveTimeSeconds` to a static value here, not in the backend (we used to default to 1/3 of the HoldTime at runtime). That way all timers are defaulted consistently, e.g. if a user does not specify them in the CR for a neighbor, it would automatically default to this when they retrieve the CR the value:
```
    virtualRouters:
    - exportPodCIDR: true
      localASN: 65001
      neighbors:
      - connectRetryTimeSeconds: 120
        holdTimeSeconds: 90
        keepAliveTimeSeconds: 30
        peerASN: 65000
        peerAddress: 172.0.0.1/32
```

Otherwise `keepAliveTimeSeconds: 30` would be omitted from defaulting on the apiserver level.

rastislavs replied:
Actually, I was thinking about this already before, so let me ask this, as now it is the last chance to change if we wanted it:

Would it make sense to wrap all timers config into its own struct (that would be optional in `CiliumBGPNeighbor`), similarly as we already do for `CiliumBGPNeighborGracefulRestart`, e.g.:
```go
type CiliumBGPNeighborTimers struct {
    ConnectRetryTimeSeconds *int32 `json:"connectRetryTimeSeconds,omitempty"`
    HoldTimeSeconds         *int32 `json:"holdTimeSeconds,omitempty"`
    KeepAliveTimeSeconds    *int32 `json:"keepAliveTimeSeconds,omitempty"`
}
```
That way, timers would not be automatically defaulted in the `CiliumBGPNeighbor`, until the user configures some of them. Then all of them would be defaulted. E.g. - without configuring timers:
```go
    virtualRouters:
    - exportPodCIDR: true
      localASN: 65001
      neighbors:
      - peerASN: 65000
        peerAddress: 172.0.0.1/32
```

Once some timer is configured:
```go
    virtualRouters:
    - exportPodCIDR: true
      localASN: 65001
      neighbors:
      - peerASN: 65000
        peerAddress: 172.0.0.1/32
        timers:
          connectRetryTimeSeconds: 120
          holdTimeSeconds: 90
          keepAliveTimeSeconds: 30
```
WDYT @harsimran-pabla & @YutaroHayakawa ? I am not sure, as that would also complicate the defaulting and nil checks in the backend though (e.g. we may still need that `applyPolicyDefaults` logic discussed here: https://github.com/cilium/cilium/pull/26040/files#r1223203259)...

YutaroHayakawa replied:
I don't have a strong opinion here, but unlike graceful restart, these timer parameters exist internally anyways, so I'd prefer to always show it to users. Also, the grouping `timer` is too generic I think. For example, we have graceful restart-related timer value, but it belongs to another group.

rastislavs replied:
> Also, the grouping timer is too generic I think

good point, maybe we should leave it as it is then

harsimran-pabla replied:
No concerns about keeping timers as individual parameters.

rastislavs commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -9,6 +9,19 @@ import (
+	// DefaultBGPPeerPort defines the TCP port number of a CiliumBGPNeighbor when PeerPort is unspecified.
+	DefaultBGPPeerPort = 179
+)
+
```
What is still missing is the type change of: 
```go
LocalASN int `json:"localASN"

PeerASN int `json:"peerASN"`
```

I think these should become `int64` to allow for maximum value `4294967295` (extended 32bit ASNs).
(`uint32` would be best, but it seems unsigned integers are not really used in k8s API)

rastislavs replied:
But note that it will probably be a larger change, as ASNs are used in various data structures in the backend, where they are manipulated as `int` (which actually could be an issue on 32bit systems, if Cilium runs on those)

rastislavs replied:
Done in https://github.com/cilium/cilium/pull/26040/commits/a811886ed98a9f856105b0fa35c267f4d12be357

harsimran-pabla commented on pkg/bgpv1/agent/controller_test.go:
```diff
@@ -143,11 +139,15 @@ func TestControllerSanity(t *testing.T) {
-							n.HoldTime.Duration != 0 &&
-							n.KeepAliveTime.Duration != 0 &&
-							n.GracefulRestart.RestartTime.Duration != 0 {
+							*n.PeerPort == v2alpha1api.DefaultBGPPeerPort &&
```
Consider refactoring this, this is bit hard to read.

harsimran-pabla commented on pkg/bgpv1/agent/controller.go:
```diff
@@ -431,62 +429,51 @@ func (c *Controller) FullWithdrawal(ctx context.Context) {
-			}
-			if n.KeepAliveTime.Duration > n.HoldTime.Duration {
+			// nil checks just in case, as everything should be defaulted in applyPolicyDefaults()
+			if n.KeepAliveTimeSeconds != nil && n.HoldTimeSeconds != nil &&
```
I think we should add validateTimers method and move this logic into it.

rastislavs replied:
Did you mean a method of `CiliumBGPNeighbor` in the `v2alpha1` package?

rastislavs replied:
I think it would make a lot of sense, done that in https://github.com/cilium/cilium/pull/26040/commits/956f08bb95b47ccaf134d2851b3ac03949d12cd0

harsimran-pabla commented on pkg/bgpv1/gobgp/server.go:
```diff
@@ -222,37 +218,38 @@ func (g *GoBGPServer) getPeerConfig(ctx context.Context, n *v2alpha1api.CiliumBG
-	peer.GracefulRestart.Enabled = n.GracefulRestart.Enabled
-	peer.GracefulRestart.RestartTime = uint32(n.GracefulRestart.RestartTime.Round(time.Second).Seconds())
-
+	if n.GracefulRestart != nil {
```
Two if statements can be merged into one - saves some indentation.

harsimran-pabla commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -154,7 +164,7 @@ type CiliumBGPVirtualRouter struct {
 	//
 	// +kubebuilder:validation:Optional
-	ExportPodCIDR bool `json:"exportPodCIDR"`
+	ExportPodCIDR *bool `json:"exportPodCIDR"`
```
Is there going to be any backward compatibility issues?

harsimran-pabla replied:
Ah read the description again - no concerns if these changes are backward compatible.

danehans replied:
Why is this field singular if a node can have multiple [PodCIDRs](https://github.com/cilium/cilium/blob/main/pkg/k8s/apis/cilium.io/client/crds/v2/ciliumnodes.yaml#L308-L314)?

From the k8s API conventions doc:
> Think twice about bool fields. Many ideas start as boolean but eventually trend towards a small set of mutually exclusive options.

Have you considered how this field could evolve where a bool type could be restrictive?

Although the field is being defaulted, I believe the `omitempty` tag should be used.

rastislavs replied:
`omitemty` fixed. As this PR aims to keep backward compatibility, renaming or changing the type other than pointer change is not possible on this field. If there will be a need for that we can do it later as a separate PR with some migration logic.

harsimran-pabla commented on pkg/bgpv1/manager/reconcile.go:
```diff
@@ -378,7 +379,7 @@ func exportPodCIDRReconciler(
 		name:      "pod CIDR",
 		component: "manager.exportPodCIDRReconciler",
-		enabled:   newc.ExportPodCIDR,
+		enabled:   pointer.BoolDeref(newc.ExportPodCIDR, false),
```
Should the default value of ExportPodCIDR be explicitly defined in API? ( No strong opinion on this )

rastislavs replied:
Yeah not sure either, but as this filed was released in a previous Cilium version, I would prefer the more conservative change to it (so, no?).

rastislavs replied:
From [Changing The API - On Compatibility](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility): "An API change is considered compatible if it: does not change existing semantics, including the semantic meaning of default values and behavior"

But then, our BGP API is still alpha, which means we in theory could do [almost any changes](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions) - but I guess we don't want it.

YutaroHayakawa replied:
Is it an incompatible change to set the default to `false` here? If not, I'd prefer to set it for consistency.

rastislavs replied:
> Is it an incompatible change to set the default to false here? 

Hmm, I think it should be fine as the default behavior remains the same. Let me change that.

rastislavs replied:
Done in https://github.com/cilium/cilium/pull/26040/commits/aa3458df112c4d33516e0e281679b3a264c8062a

YutaroHayakawa reviewed (commented):
Thank you very much for woking on this! Now our API design looks very consistent. I left some nit picking comments.

rastislavs commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -97,48 +113,47 @@ type CiliumBGPNeighbor struct {
-	// ConnectRetryTime defines the initial value for the BGP ConnectRetryTimer (RFC 4271, Section 8).
-	// The default value for the ConnectRetryTime (if empty or zero) is 120 seconds.
-	// Rounded internally to the nearest whole second.
+	EBGPMultihopTTL *int32 `json:"eBGPMultihopTTL,omitempty"`
```
Note that `EBGPMultihopTTL` is not defaulted in the API - I thought it would not make sense to default it, as it has no meaning for iBGP peers. Does it make sense or should we default it to `0` as well?

danehans replied:
Have you tested whether creating a `CiliumBGPPeeringPolicy` without this field set will cause validation to fail? If a field is optional but specifies min and max values for validation, then I believe it should be defaulted (`+kubebuilder:default=0`). Otherwise, you can drop the min validation and state "When unspecified, the default, ... is disabled".

__Note:__ This comment applies to other fields with similar properties.

rastislavs replied:
> Have you tested whether creating a CiliumBGPPeeringPolicy without this field set will cause validation to fail?

Sure, it will not cause a validation error. Validation only applies if the value is set.

YutaroHayakawa replied:
Since we indicate this option is for eBGP only, maybe setting the default to 1 (and ignoring that value for iBGP internally) makes our life a bit easier and the API looks consistent?

rastislavs replied:
> maybe setting the default to 1 (and ignoring that value for iBGP internally) makes our life a bit easier and the API looks consistent?

I like that idea, it will also address the inconsistency with the GR discussed in the other thread.

rastislavs replied:
Done in https://github.com/cilium/cilium/pull/26040/commits/ed40377c9f8db1d866745805e1a87b02f0f4bf89

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -65,19 +83,17 @@ type CiliumBGPPeeringPolicySpec struct {
 	// Enabled flag, when set enables graceful restart capability.
 	//
-	// +kubebuilder:validation:Optional
+	// +kubebuilder:validation:Required
```
`Enabled` is a bool type. From the k8s API conventions:
> Think twice about bool fields. Many ideas start as boolean but eventually trend towards a small set of mutually exclusive options.

Would it be better to use `restartTimeSeconds: 0` to indicate that graceful restart is disabled? This is the approach of multihop:
> // When empty or zero, eBGP multi-hop feature is disabled. The value is ignored for iBGP peers.

rastislavs replied:
Well, yeah I agree that we are now a bit inconsistent with regards to the `eBGPMultihopTTL`.
I remember we went this way (putting GR options into a struct) because `gracefulRestartRestartTimeSeconds` field looked weird to us :)

Thoughts @YutaroHayakawa / @harsimran-pabla ?

harsimran-pabla replied:
I think bool field for enabling/disabling graceful restart cap makes sense.

Right now, we are only exposing Restart Time, but this may change in future as we expose more configuration options described [here](https://github.com/osrg/gobgp/blob/master/api/gobgp.pb.go#L7553-L7563)

We should not tie in enabling/disabling to restart time - going back from it will be a breaking change in API.

rastislavs replied:
Good points @harsimran-pabla . At the same time I don't think that `eBGPMultihopTTL` would need its own struct and `enabled` flag similar to GR - as there are no other possible options.

YutaroHayakawa replied:
Agree to the conclusion of Pabla and Rastislav here 👍

rastislavs replied:
The inconsistency with `eBGPMultihopTTL` resolved in  https://github.com/cilium/cilium/pull/26040/commits/ed40377c9f8db1d866745805e1a87b02f0f4bf89

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -65,19 +83,17 @@ type CiliumBGPPeeringPolicySpec struct {
-	RestartTime metav1.Duration `json:"restartTime"`
+	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:default=120
+	RestartTimeSeconds *int32 `json:"restartTimeSeconds"`
```
Event hough this field is being defaulted, I believe the `omitempty` tag should be used. From the docs:
>In most cases, optional fields should also have the omitempty struct tag (the omitempty option specifies that the field should be omitted from the json encoding if the field has an empty value). However, If you want to have different logic for an optional field which is not provided vs. provided with empty values, do not use omitempty 

Thoughts?

danehans replied:
https://book.kubebuilder.io/reference/markers.html is also a good read about `+optional` vs `+kubebuilder:validation:Optional` and states:
> The most reliable way in 1.x to get +optional is omitempty.

rastislavs replied:
good catch, this should indeed have the `omitempty` tag (as well as some other optional fields that are missing it)

rastislavs replied:
Fixed in https://github.com/cilium/cilium/pull/26040/commits/df19bf07f4fb39b339257b477e17f4282c28976b

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -97,48 +113,47 @@ type CiliumBGPNeighbor struct {
-	// HoldTime defines the initial value for the BGP HoldTimer (RFC 4271, Section 4.2).
-	// The default value for the HoldTime (if empty or zero) is 90 seconds.
-	// Rounded internally to the nearest whole second. Updating this value will cause a session reset.
+	// +kubebuilder:validation:Minimum=1
```
A maximum value should be considered.

danehans replied:
Does updating this value cause a session reset?

rastislavs replied:
> Does updating this value cause a session reset?

No, this one is applied immediately without the need for session reset.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -97,48 +113,47 @@ type CiliumBGPNeighbor struct {
-	// KeepaliveTime defines the initial value for the BGP KeepaliveTimer (RFC 4271, Section 8).
-	// The default value for the KeepaliveTime (if empty or zero) is 1/3 of the HoldTime.
-	// Rounded internally to the nearest whole second. Updating this value will cause a session reset.
+	// +kubebuilder:validation:Minimum=3
```
A maximum value should be considered.

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -97,48 +113,47 @@ type CiliumBGPNeighbor struct {
 	// +kubebuilder:validation:Optional
-	// +kubebuilder:validation:Format=duration
-	KeepAliveTime metav1.Duration `json:"keepAliveTime,omitempty"`
+	// +kubebuilder:validation:Minimum=1
```
A maximum value should be considered.

danehans commented on pkg/bgpv1/agent/annotations.go:
```diff
@@ -79,7 +79,7 @@ func (e ErrAttrib) Error() string {
 type Attributes struct {
 	// The local ASN of the virtual router these Attributes targets.
-	ASN int
+	ASN int64
```
Should `LocalPort` be an `int32` like `PeerPort`?

rastislavs replied:
Not sure, it is not part of the CRD API  - it is just our internal type - so it does not have to. Only if that would be our preference.

rastislavs replied:
Actually makes a lot of sense, as it is `int32` in the `BGPGlobal` API as well. Consolidated in https://github.com/cilium/cilium/pull/26040/commits/68b5271b792f9b9f93e2800d55d1c777cf400e18

danehans commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -168,3 +184,60 @@ type CiliumBGPVirtualRouter struct {
+// Default applies default values on the CiliumBGPPeeringPolicy.
+// This is normally done by kube-apiserver for fields with explicit static defaults,
+// the main use of this method is to avoid the need for nil-checks in the controller code.
+func (p *CiliumBGPPeeringPolicy) Default() {
```
Thoughts on naming `SetDefaults()` to make it explicit for the caller?

rastislavs replied:
Personally I do not have any preference between the two - what others think about it?

harsimran-pabla replied:
+1 on SetDefaults() makes it clear that receivers' values may change.

rastislavs replied:
Actually, you changed my mind :) `SetDefaults()` will be better, as it is in fact even mutable.

danehans commented on pkg/bgpv1/agent/controller_test.go:
```diff
@@ -139,23 +137,24 @@ func TestControllerSanity(t *testing.T) {
+							*n.HoldTimeSeconds != v2alpha1api.DefaultBGPHoldTimeSeconds ||
+							*n.KeepAliveTimeSeconds != v2alpha1api.DefaultBGPKeepAliveTimeSeconds ||
+							*n.GracefulRestart.RestartTimeSeconds != v2alpha1api.DefaultBGPGRRestartTimeSeconds {
+							t.Fatalf("policy: %v not defaulted properly (invalid value)", p)
```
Thoughts on having a func that checks the defaults of a neighbor and lives by `Defaults()` to make it easier to maintain the list of defaults over time?

rastislavs replied:
Hmm, I'm not sure about that - this is a unit test, something very internal. The API package is on the other hand something very external, and such a function would not have any practical use other than this test. But at the same time I understand the inconvenience.. Would like to know what the others think about it :)

harsimran-pabla replied:
Two comments regarding this 
1. Do you think this test should move into bgpp_types_test.go? Since, we moved SetDefaults() method there, shouldn't the test to validate that Defaults are set correctly be moved within that pkg? This should make it easier to maintain, as both code and tests live in the same pkg ( harder to miss ).
2. I tend to agree with @rastislavs  on not adding a new method externally for something which is used solely in the test.

rastislavs replied:
> Do you think this test should move into bgpp_types_test.go?

Yeah, that would make sense - with one disadvantage: it would not test that `SetDefaults()` has been indeed called during the policy reconciliation in the agent.

harsimran-pabla replied:
I see, why not split the test into two parts? 

- Test_SetDefaults (in bgpp_types_test.go): Validates if values are correctly set to expected defaults 
- Test_Policy (in controller_test.go): Validates there are no nil values. 

No strong opinion, just bouncing ideas. :)

rastislavs replied:
In https://github.com/cilium/cilium/pull/26040/commits/bc81751823d7d638426726f47f3ca861c786a081 I added unit tests to the API package and removed checks for default values in the controller unit test.

danehans commented on pkg/bgpv1/gobgp/server.go:
```diff
@@ -22,8 +21,8 @@ const (
 
-	// defaultIdleHoldTimeAfterReset defines time BGP session will stay idle after neighbor reset.
-	defaultIdleHoldTimeAfterReset = 5 * time.Second
+	// defaultIdleHoldTimeAfterResetSeconds defines time BGP session will stay idle after neighbor reset.
```
nit: Not sure if this level of verbosity is needed for a local var.

rastislavs replied:
any preference for a better name?

harsimran-pabla replied:
I think we can drop `default` from the prefix since it is not really changeable. I don't have any strong preference for this.

danehans commented on pkg/bgpv1/manager/workdiff.go:
```diff
@@ -16,31 +16,31 @@ import (
 	// Local ASNs which BgpServers must be instantiated, configured,
 	// and added to the manager. Intended key for `seen` map.
-	register []int
+	register []int64
```
From the k8s api conventions:
> All public integer fields MUST use the Go (u)int32 or Go (u)int64 types, not (u)int (which is ambiguous depending on target platform). Internal types may use (u)int.

Since this is an internal type, should we continue using `int`?

rastislavs replied:
I don't think so, it could cause issues on 32-bit systems - think of storing max 32-bit ASN (`4294967295`) in a 32-bit signed integer.

rastislavs committed:
bgpv1: Add unit tests for BGP CRD types

Add unit tests to newly added SetDefaults() and Validate()
methods of the BGP CRD types. Do not perform detailed
defaulting / validation checks in unit tests of other packages,
to have that testing logic co-located within a single package.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

youngnick commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -97,48 +116,51 @@ type CiliumBGPNeighbor struct {
+	// +kubebuilder:default=30
+	KeepAliveTimeSeconds *int32 `json:"keepAliveTimeSeconds,omitempty"`
 	// GracefulRestart defines graceful restart parameters which are negotiated
 	// with this neighbor.
```
I'd suggest adding in something here about how Graceful Restart is disabled if this is empty or nil (which seems to be the case AFAICT).

This does mean that:
```yaml
    gracefulRestart:
      enabled: true
```
is the simplest way to enable this on input, which is a little verbose, but given that it means the timeout values will then be included in the output, I think it's well worth it.

youngnick commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -97,48 +116,51 @@ type CiliumBGPNeighbor struct {
 	// EBGPMultihopTTL controls the multi-hop feature for eBGP peers.
 	// Its value defines the Time To Live (TTL) value used in BGP packets sent to the neighbor.
-	// When empty or zero, eBGP multi-hop feature is disabled. The value is ignored for iBGP peers.
+	// The value 1 implies that eBGP multi-hop feature is disabled. This field is ignored for iBGP peers.
```
```suggestion
	// The value 1 implies that eBGP multi-hop feature is disabled (because only a single hop is allowed). This field is ignored for iBGP peers.
```

Small clarification request?

youngnick commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -149,22 +171,80 @@ type CiliumBGPVirtualRouter struct {
+	}
+}
+
+// SetDefaults applies default values on the CiliumBGPNeighbor.
```
This is great, because it is possible (using `kubectl apply --force` or similar) to bypass the validation checks. Nice work.

youngnick reviewed (approved):
LGTM with a few small comments, very nice work. Love seeing people run CRDs through the API conventions.

harsimran-pabla commented on pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go:
```diff
@@ -97,48 +116,51 @@ type CiliumBGPNeighbor struct {
 	//
 	// +kubebuilder:validation:Optional
-	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Minimum=1
```
Do we need to update Documentation as well regarding this change?

rastislavs replied:
Yes - not only that change, but I will do that in a separate PR to not postpone this PR any more: https://github.com/cilium/cilium/pull/26196

aanm reviewed (changes requested):
I did an overall review on the PR. It looks that pkg/k8s/apis/cilium.io/v2alpha1/bgpp_types.go got a lot of modifications, did we consider on using `v2alpha2` with a new CRD? If not, is there any upgrade path from users using v2alpha1 bgp?

aanm reviewed (approved):
Thanks for the clarification
